include/hipSYCL/sycl/detail/device_array.hpp

```diff
@@ -118,7 +87,7 @@ struct device_array
   HIPSYCL_UNIVERSAL_TARGET
   bool operator==(const device_array& other) const noexcept
   {
-    return _data == other._data;
+    return std::equal(_data, _data + N, other._data);
```

Comment:
Can we please also not use STL algorithms? It is not guaranteed that this works in device code.. For example, a simple `assert()` inside the `std::equal` implementation can break things because it might do things that require C++ standard library support, such as calling `std::terminate()`.
In general, whether STL stuff works in device code is highly dependent on the STL implementation.

Reply:
Replaced with an explicit loop.